File: android/src/main/kotlin/live/hms/hmssdk_flutter/HMSRoomExtension.kt
package live.hms.hmssdk_flutter

import android.util.Log
import live.hms.video.sdk.models.HMSRoom
import live.hms.video.sdk.models.enums.HMSRoomUpdate

class HMSRoomExtension {
    companion object{
        fun toDictionary(room:HMSRoom?):HashMap<String,Any?>?{
            val hashMap=HashMap<String,Any?>()

            if (room==null)return null
            hashMap.put("id",room.roomId )
            hashMap.put("name",room.name)
            hashMap.put("meta_data","")

            val args=ArrayList<Any>()
            room.peerList.forEach {
                args.add(HMSPeerExtension.toDictionary(it)!!)
            }
            hashMap["local_peer"] = HMSPeerExtension.toDictionary(room.localPeer)
            hashMap["peers"] = args
            hashMap["rtmp_streaming_state"] = HMSStreamingState.toDictionary(room.rtmpHMSRtmpStreamingState)
            hashMap["browser_recording_state"] = HMSStreamingState.toDictionary(room.browserRecordingState)
            hashMap["server_recording_state"] = HMSStreamingState.toDictionary(room.serverRecordingState)
            hashMap["hls_streaming_state"] = HMSStreamingState.toDictionary(room.hlsStreamingState)
            hashMap["peer_count"] = room.peerCount
            hashMap["started_at"] = room.startedAt?:-1
            hashMap["session_id"] = room.sessionId
            return hashMap
        }

        fun getValueofHMSRoomUpdate(update:HMSRoomUpdate?):String?{
            if(update==null)return null

            return when(update){
                HMSRoomUpdate.ROOM_MUTED-> "room_unmuted"
                HMSRoomUpdate.ROOM_UNMUTED-> "room_muted"
                HMSRoomUpdate.SERVER_RECORDING_STATE_UPDATED-> "server_recording_state_updated"
                HMSRoomUpdate.RTMP_STREAMING_STATE_UPDATED-> "rtmp_streaming_state_updated"
                HMSRoomUpdate.HLS_STREAMING_STATE_UPDATED-> "hls_streaming_state_updated"
                HMSRoomUpdate.BROWSER_RECORDING_STATE_UPDATED-> "browser_recording_state_updated"
                HMSRoomUpdate.ROOM_NAME_UPDATED->"room_name_updated"
                HMSRoomUpdate.ROOM_PEER_COUNT_UPDATED->"room_peer_count_updated"
                else-> "defaultUpdate"
            }
        }
    }
}

File: ios/Classes/Models/HMSRoomExtension.swift
//
//  HMSRoomExtension.swift
//  hmssdk_flutter
//
//  Copyright © 2021 100ms. All rights reserved.
//

import Foundation
import HMSSDK

class  HMSRoomExtension {

    static func toDictionary(_ room: HMSRoom) -> [String: Any] {

        var dict = [String: Any]()

        if let roomID = room.roomID {
            dict["id"] = roomID
        }

        if let name = room.name {
            dict["name"] = name
        }

        if let data = room.metaData {
            dict["meta_data"] = data
        }

        if let sessionId = room.sessionID {
            dict["session_id"] = sessionId
        }

        if let peerCount = room.peerCount {
            dict["peer_count"] = peerCount
        }

        var peers = [[String: Any]]()
        room.peers.forEach { peers.append(HMSPeerExtension.toDictionary($0)) }
        dict["peers"] = peers

        dict["rtmp_streaming_state"] = HMSStreamingStateExtension.toDictionary(rtmp: room.rtmpStreamingState)

        dict["browser_recording_state"] = HMSStreamingStateExtension.toDictionary(browser: room.browserRecordingState)

        dict["server_recording_state"] = HMSStreamingStateExtension.toDictionary(server: room.serverRecordingState)

        dict["hls_streaming_state"] = HMSStreamingStateExtension.toDictionary(hlsStreaming: room.hlsStreamingState)

        dict["hls_recording_state"] = HMSStreamingStateExtension.toDictionary(hlsRecording: room.hlsRecordingState)

        return dict
    }

    static func getValueOf(_ update: HMSRoomUpdate) -> String {
        switch update {
        case .roomTypeChanged:
            return "room_type_changed"
        case .browserRecordingStateUpdated:
            return "browser_recording_state_updated"
        case .rtmpStreamingStateUpdated:
            return "rtmp_streaming_state_updated"
        case .serverRecordingStateUpdated:
            return "server_recording_state_updated"
        case .hlsStreamingStateUpdated:
            return "hls_streaming_state_updated"
        case .hlsRecordingStateUpdated:
            return "hls_recording_state_updated"
        case .metaDataUpdated:
            return "room_peer_count_updated"
        default:
            return "unknown_update"
        }
    }
}
